src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal interface ISourceFileInfoProvider
+    {
+        /// <summary>
+        /// return SourceFileInfo for the given full path, if this provide doesn't handle the given file
+        /// return null.
+        /// </summary>
+        Task<SourceFileInfo> GetSourceFileInfoAsync(string fullPath, CancellationToken cancellationToken);
+    }
+
+    /// <summary>
+    /// SourceFileInfo is random name I choose. we might want to chagne this to something like temporarySourceInfo,
+    /// InterimSourceInfo, GeneratedSourceInfo and etc.
+    /// 
+    /// none of info returned from this can be dynamic. it should always return same data
+    /// </summary>
+    internal abstract class SourceFileInfo
+    {
+        /// <summary>
+        /// unique id to identify this file among other files in a project. it should be unique in the project
+        /// </summary>
+        public abstract string Id { get; }
+        
+        /// <summary>
+        /// file path of this source file. since this file can be virtual (doesn't actually exist in the file system),
+        /// file path doesn't need to point to actual physical file in the file system.
+        /// 
+        /// this can be used with something like editorconfig to find out right options for this source file.
+        /// </summary>
+        public abstract string FilePath { get; }
```

Comment:
So at least for right now, we have to make sure the documents added by us in the Workspace don't actually use the filename. I assume we would provide this as null for now?

Reply:
I am not sure, I was expecting something like for

c:\aaa\bbbg.cshtml you would add something like
c:\aaa\bbbg.cshtml.generated.cs or something like that. which doesn't actually exist in file system. but there at same directory as csthml and has cs extension, so that editorconfig work?

Reply:
So we could definitely provide the name of what the generated file would be during the build, I have some questions about what that would *do* - but it's not a blocker for us.

Reply:
giving it null will work as well. it won't crash.

Reply:
so, if we want to support analyzers and fixer stuff in future in razor file. we need to make it work with editorconfig. so, if we don't let it have a fake file path, then, we probably need to add some kind of abstraction so that it gets editorconfig from something for non-source file path. and user is expected to add cs/vb/dotnet code styles on editorconfig for cshtml/vbhtml?